### ISSUE TITLE
Integrate metadata cleanly into distributed software...

### ISSUE THREAD

**Issue by JustinCappos** [in-toto/in-toto#79]
We need a better plan for how to integrate link files, the layout, delegated layouts, etc. into software we distribute.  Asking the user to copy a bunch of things really doesn't cut it.  

Should there be an in-toto-bundle step which automates this (somehow)?  Should we do something different?

**Comment by lukpueh** [comment]
I totally agree. The idea has been around since the first commit, but we removed the empty `bundle` modules at some point (https://github.com/in-toto/in-toto/commit/1b344c8d39e51714ccd2d3c9dd0df09ef15c7921) since we were only dragging them along without a concrete implementation idea. Probably it's time to re-think the issue.

**Comment by JustinCappos** [comment]
One option is to have some sort of pack / unpack that happens, but this is
messy and requires certain tools to unpack.  So I think this is a
non-starter.

Another option is to put information about each final product somewhere and
to have the client retrieve it, if they know how to handle in-toto
metadata.  This has potential issues if you distribute the software via
some offline mechanism (such as a DVD for a disconnect client), but it may
be possible to work around them.  I'm not sure how we would handle the same
final product with different metadata (perhaps internal and external).
Also, how would we know whether the metadata was manipulated / omitted?  (I
guess this is a problem overall for in-toto...)  I would think the name of
the product might embed some information, such as a secure hash of the
layout+link metadata, to prevent tampering.  However, then one could tamper
with the name...

Anyways, this is an initial brain dump of rough ideas.  We should think and
discuss...

**Comment by trishankatdatadog** [comment]
Right now, to the best of my understanding, it is unclear how to distribute the top-level keys used to sign the root in-toto layout.

IMO, it's best to use TUF (with offline keys, of course) to do the job. The biggest reason to do this is as follows: in-toto needs to be able to support updating root in-toto layouts, because the security model may be upgraded iteratively. I haven't fully thought through this, but does this roughly make sense?

**Comment by lukpueh** [comment]
Here's a summary of the possibilities for shipping out in-toto metadata, we discussed last week:

1. Package layout and link metadata with the final product.
2. Provide a mapping file as part of the final product that lists URLs to fetch layout and link metadata from an external source.
3. Use TUF, in particular the `custom` field in TUF target metadata (c.f [4.5 File formats: targets.json and delegated target roles](https://github.com/theupdateframework/specification/blame/master/tuf-spec.md#L740)), which is an opaque field that could hold the dictionary representation of in-toto layout and link metadata.

**Root of trust**
In any case in-toto's root of trust, i.e. the layout signing key(s), should be served from a different location than the metadata, to separate the layout's validity from the layout.

**Discussion**
Options (1) and (2) would require extended trust in the packaging step (and its functionaries) as the step controls what metadata will be available for in-toto verification. Such a step could leave out or package arbitrary metadata to deny verification or serve outdated metadata that wrongfully passes verification.
Moreover, option (2) extends that trust even further to an additional entity, the one that is serving the metadata.

Option (3) on the other hand delegates that trust to TUF's elaborate trust model. Furthermore it maps the given metadata to a specific version of the target files and prevents from various denial of service attacks.

**Implementation**
TUF's repository tool provides an [`add_target`](https://github.com/theupdateframework/tuf/blob/develop/tuf/repository_tool.py#L1845) that accepts an optional `custom` parameter. We could use it to pass the (namespaced) in-toto metadata in a way that can be used easily by in-toto's final product verification routine, e.g.:
```python
custom = {
  "in-toto": {
    "layout": {<layout dictionary>}
    "links": {
      "<step name>" : {
        "<functionary keyid>": {<link dictionary>}
      },
      ...
    }
  }
}
```

in-toto currently tries to [load the link metadata](https://github.com/in-toto/in-toto/blob/develop/in_toto/verifylib.py#L1321-L1322) from files in the current working directory as part of the final product verification routine. We could parametrize this to look for link files elsewhere, e.g. in a specified TUF targets metadata file.

**Comment by mathisonryan** [comment]
The [Digital Bill of Materials (DBoM) open source project](https://github.com/DBOMproject) aims to solve the issue of sharing attestations around various assets in the supply chain by acting as a platform agnostic framework to share data with any of the parties. This is done by utilizing the concept of DBoM Channels which are a permissioned, policy-based store for information. These channels can have their data stored on various "repositories" which could range from traditional databases, to Distributed Ledger implementations such as Hyperledger Fabric

That being said, we the team at DBoM have found tremendous value in the in-toto project in securing the software supply chain after discussing it with @SantiagoTorres and we've tried our hand at solving this issue by integrating in-toto layout and link files into DBoM metadata entries.

We have implemented it as a wrapper around the in-toto commands. This implies that the link files, the root layout etc need not be shared along with the software package but can be retrieved securely from the DBoM channel(s) established between parties, thereby making the in-toto validation flow seamless. 

You can find the repository [here](https://github.com/DBOMproject/in-toto-dbom-wrapper). It would be quite helpful to us to know if you think that this could be of use to in-toto

**Comment by amithkk** [comment]
As an update to @mathisonryan's comment. Here is a link to a demonstration video that shows in-toto attestations being stored and retrieved across DBoM nodes: https://www.youtube.com/watch?v=uN-tdOpXtXo
